### PR TITLE
storage: Remember passphrases and export them to Anaconda

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

File: pkg/storaged/anaconda.jsx [hunk 1]
```diff
@@ -19,10 +19,108 @@
...
     // Nothing yet.
     return null;
 };
+
+export function remember_passphrase(block, passphrase) {
+    if (!client.in_anaconda_mode())
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:38]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 2]
```diff
@@ -19,10 +19,108 @@
...
+        passphrases[decode_filename(block.Device)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:42]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 3]
```diff
@@ -19,10 +19,108 @@
...
+    }
+}
+
+export function export_mount_point_mapping() {
+    if (!client.in_anaconda_mode())
+        return;
+
+    function export_info() {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:51]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 4]
```diff
@@ -19,10 +19,108 @@
...
+
+        function fstab_info(config) {
+            let dir;
+            let subvols;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:55]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 5]
```diff
@@ -19,10 +19,108 @@
...
+            for (const c of config) {
+                if (c[0] == "fstab") {
+                    const d = client.strip_mount_point_prefix(decode_filename(c[1].dir.v));
+                    if (d) {
+                        const sv = parse_subvol_from_options(decode_filename(c[1].opts.v));
+                        if (sv) {
+                            if (sv.pathname) {
+                                if (!subvols)
+                                    subvols = { };
+                                subvols[sv.pathname] = d;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:66]: These 10 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 6]
```diff
@@ -19,10 +19,108 @@
...
+                            }
+                        } else if (!dir) {
+                            dir = d;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:69]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 7]
```diff
@@ -19,10 +19,108 @@
...
+                        }
+                    }
+                }
+            }
+
+            if (dir || subvols)
+                return {
+                    type: "filesystem",
+                    dir,
+                    subvolumes: subvols
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:79]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 8]
```diff
@@ -19,10 +19,108 @@
...
+                };
+        }
+
+        function block_info(block) {
+            if (block.IdUsage == "filesystem") {
+                return fstab_info(block.Configuration);
+            } else if (block.IdUsage == "swap") {
+                return {
+                    type: "swap",
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:88]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 9]
```diff
@@ -19,10 +19,108 @@
...
+                };
+            } else if (block.IdUsage == "crypto") {
+                const cleartext_block = client.blocks_cleartext[block.path];
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:91]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 10]
```diff
@@ -19,10 +19,108 @@
...
+
+                let content_info;
+                if (cleartext_block)
+                    content_info = block_info(cleartext_block);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:95]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 11]
```diff
@@ -19,10 +19,108 @@
...
+                else {
+                    const block_crypto = client.blocks_crypto[block.path];
+                    if (block_crypto)
+                        content_info = fstab_info(block_crypto.ChildConfiguration);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:99]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 12]
```diff
@@ -19,10 +19,108 @@
...
+                }
+
+                return {
+                    type: "crypto",
+                    cleartext_device: cleartext_block && decode_filename(cleartext_block.Device),
+                    content: content_info,
+                    passphrase: passphrases[decode_filename(block.Device)],
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:106]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 13]
```diff
@@ -19,10 +19,108 @@
...
+                };
+            } else
+                return undefined;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:109]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 14]
```diff
@@ -19,10 +19,108 @@
...
+        }
+
+        const mpm = { };
+        for (const p in client.blocks) {
+            const b = client.blocks[p];
+            mpm[decode_filename(b.Device)] = block_info(b);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:115]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 15]
```diff
@@ -19,10 +19,108 @@
...
+        }
+
+        window.sessionStorage.setItem("cockpit_mount_points", JSON.stringify(mpm));
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:118]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 16]
```diff
@@ -19,10 +19,108 @@
...
+    }
+
+    try {
+        export_info();
+    } catch {
+        console.warn("Error exporting mount points");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:124]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx [hunk 17]
```diff
@@ -78,17 +79,18 @@ function clevis_unlock(block) {
...
-        return crypto.Unlock("", {});
-    else if (passphrase_type == "clevis")
-        return clevis_unlock(block);
-    else {
+    if (passphrase) {
+        await crypto.Unlock(passphrase, {});
+        remember_passphrase(block, passphrase);
+    } else if (passphrase_type == "stored") {
+        await crypto.Unlock("", {});
+    } else if (passphrase_type == "clevis") {
```
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240201-153929-63faa098-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by KKoukiou — COMMENTED** [review]
When creating a device and then formatting it this works. However I create an encrypted partition directly the passphrase is not passed or at least not maintained inside the cockpit_passphrases. 

https://github.com/cockpit-project/cockpit/assets/14921356/d4e847ed-5705-4316-8cdc-deb1c4a494b1

**Comment by mvollmer** [comment]
> When creating a device and then formatting it this works. However I create an encrypted partition directly the passphrase is not passed or at least not maintained inside the cockpit_passphrases.

Ah, yes, I forgot about this case. I was only looking for the Unlock() method... Thanks!

**Review by KKoukiou — COMMENTED** [review]
When creating an encrypted logical volume, I see:

Error exporting mount points

Screen Shot 2024-02-01 at 18 31 43

**Comment by mvollmer** [comment]
> When creating an encrypted logical volume, I see:
> 
> Error exporting mount points

Yes, hmm, let's not hide all errors when exporting, I went too far with that. I don't want the page to crash when there is garbage in sessionStorage, but I think we should crash when there is a bug in the exporting code.

**Comment by mvollmer** [comment]
> > When creating a device and then formatting it this works. However I create an encrypted partition directly the passphrase is not passed or at least not maintained inside the cockpit_passphrases.
> 
> Ah, yes, I forgot about this case. I was only looking for the Unlock() method... Thanks!

Should be fixed.

**Comment by mvollmer** [comment]
> When creating an encrypted logical volume, I see:
> 
> Error exporting mount points

I can't reproduce this, unfortunately.

File: pkg/storaged/anaconda.jsx [hunk 18]
```diff
@@ -19,10 +19,105 @@
...
     // Nothing yet.
     return null;
 };
+
+export function remember_passphrase(block, passphrase) {
+    if (!client.in_anaconda_mode())
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:38]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 19]
```diff
@@ -19,10 +19,105 @@
...
+        passphrases[decode_filename(block.Device)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:42]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 20]
```diff
@@ -19,10 +19,105 @@
...
+
+export function export_mount_point_mapping() {
+    if (!client.in_anaconda_mode())
+        return;
+
+    let passphrases = { };
+    try {
+        passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
+    } catch {
+        console.warn("Can't read passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:54]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 21]
```diff
@@ -19,10 +19,105 @@
...
+    }
+
+    function fstab_info(config) {
+        let dir;
+        let subvols;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:59]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 22]
```diff
@@ -19,10 +19,105 @@
...
+        for (const c of config) {
+            if (c[0] == "fstab") {
+                const d = client.strip_mount_point_prefix(decode_filename(c[1].dir.v));
+                if (d) {
+                    const sv = parse_subvol_from_options(decode_filename(c[1].opts.v));
+                    if (sv) {
+                        if (sv.pathname) {
+                            if (!subvols)
+                                subvols = { };
+                            subvols[sv.pathname] = d;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:70]: These 10 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 23]
```diff
@@ -19,10 +19,105 @@
...
+                        }
+                    } else if (!dir) {
+                        dir = d;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:73]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 24]
```diff
@@ -19,10 +19,105 @@
...
+                    }
+                }
+            }
+        }
+
+        if (dir || subvols)
+            return {
+                type: "filesystem",
+                dir,
+                subvolumes: subvols
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:83]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 25]
```diff
@@ -19,10 +19,105 @@
...
+            };
+    }
+
+    function block_info(block) {
+        if (block.IdUsage == "filesystem") {
+            return fstab_info(block.Configuration);
+        } else if (block.IdUsage == "swap") {
+            return {
+                type: "swap",
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:92]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 26]
```diff
@@ -19,10 +19,105 @@
...
+            };
+        } else if (block.IdUsage == "crypto") {
+            const cleartext_block = client.blocks_cleartext[block.path];
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:95]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 27]
```diff
@@ -19,10 +19,105 @@
...
+
+            let content_info;
+            if (cleartext_block)
+                content_info = block_info(cleartext_block);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:99]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 28]
```diff
@@ -19,10 +19,105 @@
...
+            else {
+                const block_crypto = client.blocks_crypto[block.path];
+                if (block_crypto)
+                    content_info = fstab_info(block_crypto.ChildConfiguration);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:103]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 29]
```diff
@@ -19,10 +19,105 @@
...
+            }
+
+            if (content_info)
+                return {
+                    type: "crypto",
+                    cleartext_device: cleartext_block && decode_filename(cleartext_block.Device),
+                    content: content_info,
+                    passphrase: passphrases[decode_filename(block.Device)],
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:111]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 30]
```diff
@@ -19,10 +19,105 @@
...
+                };
+        }
+    }
+
+    const mpm = { };
+    for (const p in client.blocks) {
+        const b = client.blocks[p];
+        mpm[decode_filename(b.Device)] = block_info(b);
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:119]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 31]
```diff
@@ -19,10 +19,105 @@
...
+    }
+
+    window.sessionStorage.setItem("cockpit_mount_points", JSON.stringify(mpm));
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:122]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 32]
```diff
@@ -631,18 +632,26 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
...
+                        await client.mount_at(block, mount_point);
+                    }
+                    if (type == "swap" && mount_now) {
+                        const block_swap = await client.wait_for(() => block_swap_for_block(path));
+                        const block = client.blocks[block_swap.path];
+                        if (is_encrypted(vals))
+                            remember_passphrase(block, vals.passphrase);
+                        await block_swap.Start({});
+                    }
+                    if (is_encrypted(vals) && (is_filesystem(vals) || type == "swap") && !mount_now) {
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:651]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240202-090532-99f6d966-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 33]
```diff
@@ -30,6 +30,19 @@
...
+export function remember_passphrase(block, passphrase) {
+    if (!client.in_anaconda_mode())
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
+        passphrases[decode_filename(block.Device)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:42]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240205-141136-e6895fca-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx (hunk 32 above)
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:651]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240205-141136-e6895fca-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240205-141136-e6895fca-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx (hunk 33 above)
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:42]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240206-142507-3fbe954a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx (hunk 32 above)
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:651]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240206-142507-3fbe954a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240206-142507-3fbe954a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by KKoukiou — APPROVED** [review]

File: doc/anaconda.md [hunk 34]
```diff
@@ -124,6 +110,7 @@ encoded object, for example:
   },
   "/dev/sdc": {
     "type": "crypto",
+    "passphrase": "foobar",
```
Comment (mvollmer) [inline doc/anaconda.md:113]: This should be removed, this field is no longer set.

Comment (mvollmer) [reply doc/anaconda.md]: Done

**Review by martinpitt — DISMISSED** [review]
I don't put my harsh words into GitHub, but I cannot in good faith approve this. Let's *pretty please* fix that anaconda bug to not lock an unlocked LUKS device in the target system.

**Comment by mvollmer** [comment]
Oops, we need tests, forgot about that.

**Comment by martinpitt** [comment]
Ah, I missed that it skips the storing of passphrases in non-anaconda mode, this alleviates most of my concern. So I retract my veto.

File: pkg/storaged/anaconda.jsx (hunk 33 above)
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:42]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-085751-54761d20-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx (hunk 32 above)
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:651]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-085751-54761d20-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-085751-54761d20-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 35]
```diff
@@ -30,19 +30,57 @@
     return null;
 };
 
+function parse_parent_from_options(options) {
+    const parent_match = options.match(/x-parent=(?<parent>[\w\\-]+)/);
+    if (parent_match) {
+        return parent_match.groups.parent;
+    } else
+        return null;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:38]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-125230-6cf61ca9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 36]
```diff
@@ -30,19 +30,57 @@
...
+
+    if (!window.isSecureContext)
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
+        passphrases[decode_filename(block.Device)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:57]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-125230-6cf61ca9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 37]
```diff
@@ -96,8 +153,22 @@
...
     for (const p in client.blocks) {
         const b = client.blocks[p];
-        mpm[decode_filename(b.Device)] = block_info(b);
+        mpm[device_name(b)] = block_info(b);
+    }
+
+    // Add inactive logical volumes
+    for (const vg_p in client.vgroups) {
+        const vg = client.vgroups[vg_p];
+        for (const lv of client.vgroups_lvols[vg_p] || []) {
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:162]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-125230-6cf61ca9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 38]
```diff
@@ -631,18 +621,24 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
...
+                        remember_passphrase(new_block, vals.passphrase);
+
+                    if (is_filesystem(vals) && mount_now) {
+                        await client.mount_at(new_block, mount_point);
+                    }
+                    if (type == "swap" && mount_now) {
+                        const block_swap = await client.wait_for(() => block_swap_for_block(path));
+                        await block_swap.Start({});
+                    }
+                    if (is_encrypted(vals) && (is_filesystem(vals) || type == "swap") && !mount_now) {
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:638]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-125230-6cf61ca9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240207-125230-6cf61ca9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 39]
```diff
@@ -22,19 +22,57 @@
 import { decode_filename } from "./utils.js";
 import { parse_subvol_from_options } from "./btrfs/utils.jsx";
 
+function parse_parent_from_options(options) {
+    const parent_match = options.match(/x-parent=(?<parent>[\w\\-]+)/);
+    if (parent_match) {
+        return parent_match.groups.parent;
+    } else
+        return null;
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:30]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240209-085628-56676a47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 40]
```diff
@@ -22,19 +22,57 @@
...
+
+    if (!window.isSecureContext)
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
+        passphrases[decode_filename(block.Device)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:49]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240209-085628-56676a47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/anaconda.jsx [hunk 41]
```diff
@@ -88,8 +145,22 @@
...
     for (const p in client.blocks) {
         const b = client.blocks[p];
-        mpm[decode_filename(b.Device)] = block_info(b);
+        mpm[device_name(b)] = block_info(b);
+    }
+
+    // Add inactive logical volumes
+    for (const vg_p in client.vgroups) {
+        const vg = client.vgroups[vg_p];
+        for (const lv of client.vgroups_lvols[vg_p] || []) {
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:154]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240209-085628-56676a47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240209-085628-56676a47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by KKoukiou — APPROVED** [review]
As mentioned on matrix.
* Please merge this asap, tests are currently passing with this and my open cockpit-storage integration PR
 * create a ticket in COCKPIT component in Jira to remove this code which will be blocked on https://issues.redhat.com/browse/INSTALLER-3898
 * I already started working on https://issues.redhat.com/browse/INSTALLER-3898 and it looks very promising - even for this f40 release

 Merging prematurely this - even though it's not final solution - will make our collaboration more seemless as we can develop on top of existing tests to at least not break the functionality

File: pkg/storaged/anaconda.jsx [hunk 42]
```diff
@@ -34,15 +34,31 @@
...
-                decode_filename(block.PreferredDevice));
+    if (!window.isSecureContext)
+        return;
+
+    try {
+        const passphrases = JSON.parse(window.sessionStorage.getItem("cockpit_passphrases")) || { };
+        passphrases[device_name(block)] = passphrase;
+        window.sessionStorage.setItem("cockpit_passphrases", JSON.stringify(passphrases));
+    } catch {
+        console.warn("Can't record passphrases");
```
Comment (cockpituous) [inline pkg/storaged/anaconda.jsx:55]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240212-163533-e482442c-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 17 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19924-20240212-163533-e482442c-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Pull request merged**